File: src/test/java/com/jacuum/engine/MemorySessionsTest.java
package com.jacuum.engine;

import com.jacuum.algo.*;
import com.jacuum.map.*;
import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.*;

class MemorySessionsTest {

    // 3×3 all-floor map, start at (1,1)
    private static GameMap smallMap() {
        boolean[][] f = new boolean[3][3];
        for (int y = 0; y < 3; y++) for (int x = 0; x < 3; x++) f[y][x] = true;
        return new GeneratedMap("test", SizePreset.TINY, f, 1, 1);
    }

    @Test void openCreatesSessionInSetupState() throws Exception {
        Sessions sessions = new MemorySessions(null, null, 50); // no messaging in unit test
        String id = sessions.open(smallMap(), "RandomAlgo", "Alice", "🤖", 100);
        assertThat(id).isNotBlank();
        SessionView view = sessions.view(id);
        assertThat(view.status()).isEqualTo(RunStatus.SETUP);
        assertThat(view.robotX()).isEqualTo(1);
        assertThat(view.robotY()).isEqualTo(1);
        assertThat(view.score()).isEqualTo(0);
        assertThat(view.totalFloor()).isEqualTo(9);
    }

    @Test void viewThrowsForUnknownId() {
        Sessions sessions = new MemorySessions(null, null, 50);
        assertThatThrownBy(() -> sessions.view("nope"))
            .isInstanceOf(Exception.class);
    }
}

File: src/test/kotlin/com/jacuum/engine/MemorySessionsTest.kt
package com.jacuum.engine

import com.jacuum.map.GeneratedMap
import com.jacuum.map.SizePreset
import org.assertj.core.api.Assertions.assertThat
import org.assertj.core.api.Assertions.assertThatThrownBy
import org.junit.jupiter.api.Test

class MemorySessionsTest {

    private fun smallMap(): com.jacuum.map.GameMap {
        val f = Array(3) { BooleanArray(3) { true } }
        return GeneratedMap("test", SizePreset.TINY, f, 1, 1)
    }

    @Test fun openCreatesSessionInSetupState() {
        val sessions = MemorySessions(SilentMessaging(), null, 50)
        val id = sessions.open(smallMap(), "RandomAlgo", "Alice", "\uD83E\uDD16", 100)
        assertThat(id).isNotBlank()
        val view = sessions.view(id)
        assertThat(view.status).isEqualTo(RunStatus.SETUP)
        assertThat(view.robotX).isEqualTo(1)
        assertThat(view.robotY).isEqualTo(1)
        assertThat(view.score).isEqualTo(0)
        assertThat(view.totalFloor).isEqualTo(9)
    }

    @Test fun viewThrowsForUnknownId() {
        val sessions = MemorySessions(SilentMessaging(), null, 50)
        assertThatThrownBy { sessions.view("nope") }.isInstanceOf(Exception::class.java)
    }

    @Test fun sessionCapRejected() {
        val sessions = MemorySessions(SilentMessaging(), null, 2)
        sessions.open(smallMap(), "a", "Alice", "\uD83E\uDD16", 100)
        sessions.open(smallMap(), "a", "Bob", "\uD83E\uDD16", 100)
        assertThatThrownBy { sessions.open(smallMap(), "a", "Charlie", "\uD83E\uDD16", 100) }
            .isInstanceOf(Exception::class.java)
            .hasMessageContaining("Session cap reached")
    }

    @Test fun pauseRequiresRunningState() {
        val sessions = MemorySessions(SilentMessaging(), null, 50)
        val id = sessions.open(smallMap(), "a", "Alice", "\uD83E\uDD16", 100)
        assertThatThrownBy { sessions.pause(id) }.isInstanceOf(Exception::class.java)
    }

    @Test fun resumeRequiresPausedState() {
        val sessions = MemorySessions(SilentMessaging(), null, 50)
        val id = sessions.open(smallMap(), "a", "Alice", "\uD83E\uDD16", 100)
        assertThatThrownBy { sessions.resume(id) }.isInstanceOf(Exception::class.java)
    }
}
